### ISSUE TITLE
kernel.js and extension mechanism?

### ISSUE THREAD

**Issue by flying-sheep** [jupyterlab/jupyterlab#3010]
hi, i’d like to implement kernel-specific functionality for both jupyterlab and jupyter notebook.

the latter has `kernel.js`, but i don’t see a way how to extend the jupyterlab notebook (and maybe also console).

specifically, i want to add

- two editor shortcuts that insert language-specific code at the cursor
- a shortcut that pops up help (for the object below the cursor) in a pager. i know that inspect_request is used for this

**Comment by blink1073** [comment]
Hi @flying-sheep, unfortunately there will not be a `kernel.js` equivalent in JupyterLab, because extensions have to be built into the application itself.  We don't yet have an exposed method to add editor shortcuts to the notebook.

**Comment by blink1073** [comment]
Do you mind opening a separate issue for the ability to add shortcuts?  Once that is in place, an extension could be written similar to the example [notebook extension](https://github.com/jupyterlab/jupyterlab/blob/master/docs/notebook.md#adding-a-button-to-the-toolbar) that adds/removes the editor shortcuts.

**Comment by flying-sheep** [comment]
sure!  #3022

**Comment by soli** [comment]
Hi, sorry if my question is only partially related (my search-fu fails me miserably here) but we're also providing a (wrapper) kernel for Jupyter and I'd like to provide JupyterLab support. Most of the things work out-of-the-box (now that bokeh is properly handling JupyterLab) except what we used to do in `kernel.js`: defining a CodeMirror SimpleMode*…

Am I right that I should now create a jupyterlab extension for this? Could you point me to existing ones (for this specific aspect, i.e. extending CodeMirror) ? Otherwise would starting from `codemirror-ipython.ts` make sense ?
Since adding language support via `kernel.js` was explicitly documented in the Jupyter kernel docs, it might be useful to add a link to some help in porting what exists for Jupyter to JupyterLab (the current JupyterLab docs do not seem to consider that one might want to write a kernel for JupyterLab at all unfortunately. "Everything is an extension"… except kernels 😉)

*: we prefer, at that point, not to include our language support into the main CodeMirror repo… that might change later on, but for now I cannot rely on that 😢

**Comment by SpencerPark** [comment]
Also ran into an issue around the `kernel.js` support not being available in jupyterlab. Specifically for creating a kernel for a new language in which I was also distributing the codemirror mode via the `kernel.js`.

This mechanism made a lot of sense to me, ship the syntax highlighter with the kernel. Is an extension still the only way to do this?

**Comment by soli** [comment]
I'm afraid so… some people manage to do this as an extension, like [here](https://github.com/vatlab/jupyterlab-sos/blob/master/src/codemirror-sos.ts), but this seems a huge lot of boilerplate just to ship some simple JS with a given kernel (and also creates an extension which doesn't make any sense without the corresponding kernel).

**Comment by jasongrout** [comment]
One issue is that it almost never is just some simple standalone JS. Often that JS needs some dependencies, which pull in other dependencies (and how are all those packaged up?), then you have to deal with conflict resolution and deduplication with existing packages, etc.

That said, it would be great if we had a much lighter way of loading small extensions on the fly. For example, I'd love to be able to have quick one-off extensions developed and loaded solely from within JLab (e.g., here's a JS file that contains the extension code that only depends on packages already existing in jlab - right click and load it as an extension).

**Comment by soli** [comment]
Well, kernel.js didn't help for the complicated scenario you describe. What it was good for, on the other hand, is simple standalone JS that is just for the kernel.

The typical example is CodeMirror mode. This very example was even in the docs of Jupyter not-lab…

For more complex stuff, I agree that extensions are more convenient, but for stuff that is really kernel-dependent, it does not really make sense to me. Until kernels also become extensions, which is probably when I'll stop making kernels ;)

**Comment by SpencerPark** [comment]
I agree, that would be great! Like I mentioned previously, I am mostly interested in being able to ship a syntax highlighter. As such I would be fine with something even more restricted like a `codemirror.js` in the kernel directory so that it decouples jupyterlab from codemirror if that is a concern. The point being that this type of "extension" really only makes sense when a specific kernel is running and should be more tightly coupled to it rather than a standalone component.

**Comment by jasongrout** [comment]
> The typical example is CodeMirror mode.

Even that will be require some things from the system, including codemirror, so even that scenario involves some of the complications I mentioned. I think it's doable, but it could be tricky and requires some thought in its design. If someone wants to tackle how to provide lightweight extensions, or even if perhaps there is a new lightweight extension framework just for codemirror modes (sort of like we now have a simplified rendermime extension), we can help them get started.

**Comment by soli** [comment]
Could you point me/us to these simplified rendermime extensions? I have no idea when I'll have some time to look this up, but it's almost Christmas, we can hope for a miracle XD

**Comment by SpencerPark** [comment]
I believe https://jupyterlab.readthedocs.io/en/stable/developer/extension_dev.html#mime-renderer-extensions is what is being referred to here with an emphasis on the `rendermime-interfaces` package. I think that we can put together something similar for codemirror modes.

I've found https://github.com/jupyterlab/jupyterlab/blob/018dc601ecfb222794af6dcc032e56376a46764b/packages/help-extension/src/index.tsx#L211 which makes me believe getting access to the script should not be an issue for the front-end (although I am a little confused as to why it is a map and not a function other than for making it serializable?). 

Also [codemirror modes](https://codemirror.net/doc/manual.html#modeapi) already encourage very localized definitions with the single entry point `defineMode` which could be the same interface we expose in this lightweight codemirror extension. A kernel may specify a `codemirror-mode.js` which exports a name, modetype, and mode. Jupyterlab can define (or undefined?) as necessary to ensure that cells are properly highlighted.

The _modetype_ could be either `'standard'`, `'simple'` (https://codemirror.net/demo/simplemode.html), `'extension'`, or `'overlay'` to cover all the types.

Is the codemirror instance global with respect to the window or on an off chance that JupyterLab has worked some magic, local to a notebook?

Lastly another thing we could do is use the mime-type and codemirror mode (or name) from the `language_info` instead of letting it be configured separately which would restrict kernels to only defining modes for themselves.

Of course this is a lot of (somewhat incoherent) ideas but I wanted to put them here to hopefully get something started. I would be happy to help work on this as well.

**Comment by jamesdbrock** [comment]
Any update on this? Maybe the new JupyterLab v3 pre-built-extension feature provides a path for bundling a labextension with a kernel?

I want the same thing everyone here wants: an easy way to provide a custom Codemirror mode for syntax highlighting for a specific kernel without requiring the user to “install” a labextension.

**Comment by krassowski** [comment]
Yes, CodeMirror modes can be installed as pre-built extensions and distributed via pip, conda or other package manager. One example is https://github.com/MarketSquare/jupyterlab_robotmode. It should be also possible to bundle a kernel and code mirror extension into a single package so users don't have to install it manually. As for pre-built extensions, the mode has to be defined roughly as in https://github.com/jupyterlab/jupyterlab/issues/10095#issuecomment-819889990.

If you believe it should be done differently to help kernel maintainers, we may have an opportunity to re-think this during migration to CodeMirror 6 (https://github.com/jupyterlab/jupyterlab/issues/10370) as it will come with improved highlighting modes system which will also likely drop `ICodeEditor` abstraction (https://github.com/jupyterlab/jupyterlab/issues/10380).
